### PR TITLE
feat: nexus gateway contract route messages to nexus only if they have not been routed yet

### PR DESCRIPTION
**Pull request by fish-sammy**
## Todos

- [x] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: contracts/nexus-gateway/src/state.rs [hunk 1]
```diff
@@ -23,14 +31,29 @@ impl<'a> GatewayStore<'a> {
...
+        CONFIG
+            .save(self.storage, &config)
+            .change_context(ContractError::StoreFailure)
     }
 
-    fn load_config(&self) -> Config {
+    fn load_config(&self) -> Result<Config> {
         CONFIG
             .load(self.storage)
-            .expect("config must be set during contract instantiation")
```
Comment (cjcobb23) [inline contracts/nexus-gateway/src/state.rs:33]: why this change? Feel like this should never happen.

File: contracts/nexus-gateway/src/contract/execute.rs [hunk 2]
```diff
@@ -1,17 +1,54 @@
...
+        for msg in &msgs {
+            self.store.set_message_routed(&msg.cc_id)?;
+        }
+
+        let msgs = msgs
+            .into_iter()
+            .map(TryInto::try_into)
+            .collect::<Result<Vec<nexus::Message>>>()?;
+
+        Ok(Response::new().add_messages(msgs))
```
Comment (cjcobb23) [inline contracts/nexus-gateway/src/contract/execute.rs:50]: This message will definitely error if the nexus fails to route the message right? Just want to make sure a message isn't set as routed here, but then not actually routed in the nexus.

Comment (fish-sammy) [reply contracts/nexus-gateway/src/contract/execute.rs:50]: yes

File: contracts/nexus-gateway/src/state.rs (hunk 1 above)
Comment (fish-sammy) [reply contracts/nexus-gateway/src/state.rs:33]: This could happen if deserialization fails. For example, we have some custom deserializer implemented for the Config where we do some custom validation.

Comment (cjcobb23) [reply contracts/nexus-gateway/src/state.rs:33]: yeah but shouldn't we panic in that situation?

File: contracts/nexus-gateway/src/contract/execute.rs (hunk 2 above)
Comment (cjcobb23) [reply contracts/nexus-gateway/src/contract/execute.rs:50]: This actually makes me wonder, why don't we track the messages in nexus, instead of here? Since nexus has the restriction of only once routing, why not just enforce that there?

File: contracts/nexus-gateway/src/contract/execute.rs [hunk 3]
```diff
@@ -1,17 +1,54 @@
...
+        mut self,
+        sender: Addr,
+        msgs: Vec<Message>,
+    ) -> Result<Response<nexus::Message>> {
+        match sender {
+            sender if sender == self.config.nexus => todo!(),
+            sender if sender == self.config.router => self
+                .route_to_nexus(msgs)
+                .change_context(ContractError::RouteToNexus),
+            _ => Err(ContractError::Unauthorized)?,
```
Comment (cgorenflo) [inline contracts/nexus-gateway/src/contract/execute.rs:27]: it's sad that this has to be a conditional match instead of a direct equality comparison, but I still prefer this match pattern over an `if.. else if` chain

Comment (cgorenflo) [inline contracts/nexus-gateway/src/contract/execute.rs:27]: I think I prefer `Err(report!(ContractError::Unauthorized))` or `bail!(ContractError::Unauthorized)` over the use of `?` here

File: contracts/nexus-gateway/src/contract/execute.rs [hunk 4]
```diff
@@ -1,17 +1,54 @@
...
+    pub fn route_messages(
+        mut self,
+        sender: Addr,
+        msgs: Vec<Message>,
+    ) -> Result<Response<nexus::Message>> {
+        match sender {
+            sender if sender == self.config.nexus => todo!(),
+            sender if sender == self.config.router => self
+                .route_to_nexus(msgs)
+                .change_context(ContractError::RouteToNexus),
```
Comment (cgorenflo) [inline contracts/nexus-gateway/src/contract/execute.rs:26]: consider moving the context switch into `route_to_nexus`

File: contracts/nexus-gateway/src/contract/execute.rs [hunk 5]
```diff
@@ -1,17 +1,54 @@
...
+                Ok(false) => Some(Ok(msg)),
+                Err(err) => Some(Err(err)),
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        for msg in &msgs {
+            self.store.set_message_routed(&msg.cc_id)?;
+        }
+
+        let msgs = msgs
```
Comment (cgorenflo) [inline contracts/nexus-gateway/src/contract/execute.rs:45]: use `inspect` to make all of this a single expression

File: contracts/nexus-gateway/src/contract/execute.rs [hunk 6]
```diff
@@ -20,18 +57,115 @@ mod test {
 
     use connection_router::state::{CrossChainId, Message};
 
+    use crate::state::{Config, MockStore};
+
     use super::*;
 
+    #[test]
+    fn test_route_messages_unauthorized() {
```
Comment (cgorenflo) [inline contracts/nexus-gateway/src/contract/execute.rs:65]: we don't have to start test functions with `test_`. They already have the `#[test]` attribute and from what I've seen so far it seems to be convention to directly describe the test, e.g. `do_not_route_messages_from_unauthorized_sender`

File: contracts/nexus-gateway/src/contract/execute.rs [hunk 7]
```diff
@@ -20,18 +57,115 @@ mod test {
...
+        let store = MockStore::new();
+        let config = Config {
+            nexus: Addr::unchecked("nexus"),
+            router: Addr::unchecked("router"),
+        };
+        let contract = Contract { store, config };
+
+        let res = contract.route_messages(Addr::unchecked("unauthorized"), vec![]);
+
+        assert!(res.is_err_and(|err| matches!(err.current_context(), ContractError::Unauthorized)));
```
Comment (cgorenflo) [inline contracts/nexus-gateway/src/contract/execute.rs:75]: nice! I like the use of `is_err_and`

File: contracts/nexus-gateway/src/contract/execute.rs [hunk 8]
```diff
@@ -20,18 +57,115 @@ mod test {
...
+                destination_address: "0x3fC91A3afd70395Cd496C647d5a6CC9D4B2b7FAD"
+                    .parse()
+                    .unwrap(),
+                destination_chain: "destinationChain".parse().unwrap(),
+                payload_hash: decode(
+                    "bb9b5566c2f4876863333e481f4698350154259ffe6226e283b16ce18a64bcf1",
+                )
+                .unwrap()
+                .into(),
+            },
```
Comment (cgorenflo) [inline contracts/nexus-gateway/src/contract/execute.rs:149]: can you make this more compact? The exact content of messages doesn't really matter for these tests. (E.g. use shorter hexes so the parse().unwrap() doesn't wrap or extract the message creation into another function

File: contracts/nexus-gateway/src/state.rs (hunk 1 above)
Comment (cgorenflo) [reply contracts/nexus-gateway/src/state.rs:33]: panic if the error is due to a programming error, Err() if it is due to input data (or something circumstantial like unavailable services).
So if the some validation of the config fails that should be an Err

Comment (cjcobb23) [reply contracts/nexus-gateway/src/state.rs:33]: But shouldn't we not save a config that can't be deserialized? Which would mean if we get here, then we saved a config that we shouldn't have saved, hence it would be a programming error.

My point is, I don't see how we can get to this state, where the config can't be loaded. It seems like a totally invalid contract state to me. I also don't know what can even be done on the contract if the config can't be loaded, the whole contract is basically dead at this point.

Comment (cgorenflo) [reply contracts/nexus-gateway/src/state.rs:33]: at that point you are baking knowledge about which data came from the outside and which was created internally into the store. What if the config was never (de)serialized before? It's still very easy to convert this Err into a panic at a higher level were we have explicit knowledge about where the struct came from

Comment (cgorenflo) [reply contracts/nexus-gateway/src/state.rs:33]: in conclusion: 
- should the contract panic if the config can't be loaded? Most likely, yes
- should the store know in which cases it's safe to panic vs returning an Err? I don't think so

File: contracts/nexus-gateway/src/contract/execute.rs (hunk 2 above)
Comment (cgorenflo) [reply contracts/nexus-gateway/src/contract/execute.rs:50]: good point!

Comment (fish-sammy) [reply contracts/nexus-gateway/src/contract/execute.rs:50]: This contract is an extension of nexus as the name suggests. The code here is more upgradable than nexus.

File: contracts/nexus-gateway/src/contract/execute.rs (hunk 3 above)
Comment (fish-sammy) [reply contracts/nexus-gateway/src/contract/execute.rs:27]: sure

File: contracts/nexus-gateway/src/state.rs (hunk 1 above)
Comment (fish-sammy) [reply contracts/nexus-gateway/src/state.rs:33]: Same to me. It should be the contract instead of the store to determine when to panic.

File: contracts/nexus-gateway/src/contract/execute.rs (hunk 4 above)
Comment (fish-sammy) [reply contracts/nexus-gateway/src/contract/execute.rs:26]: I like it to be here cuz `route_to_nexus` reports the specifics about why `RouteToNexus` occurs.

File: contracts/nexus-gateway/src/contract/execute.rs (hunk 5 above)
Comment (fish-sammy) [reply contracts/nexus-gateway/src/contract/execute.rs:45]: Can't do `inspect` cuz `set_message_routed` returns a Result.

File: contracts/nexus-gateway/src/contract/execute.rs (hunk 8 above)
Comment (fish-sammy) [reply contracts/nexus-gateway/src/contract/execute.rs:149]: sure

File: contracts/nexus-gateway/src/contract/execute.rs [hunk 9]
```diff
@@ -20,54 +56,135 @@ mod test {
...
+        ];
+        let res = contract.route_messages(Addr::unchecked("router"), msgs);
+
+        assert!(res.is_ok_and(|res| res.messages.len() == 2));
     }
 
     #[test]
-    fn test_route_to_nexus_with_msgs() {
+    fn route_to_nexus_with_msgs_that_have_been_routed() {
```
Comment (cjcobb23) [inline contracts/nexus-gateway/src/contract/execute.rs:144]: might be good to test a mix of messages that have been routed and some that haven't

**Review by cjcobb23 — APPROVED** [review]

**Review by cgorenflo — APPROVED** [review]
